### PR TITLE
Support modified source files for refresh incremental mode

### LINKED ISSUE #182: Improve RefreshIncremental logic to take last modified time of files and file size into account.
### Describe the issue

RefreshIncremental logic doesn't take into account file size and timestamp for comparing files. Improve this logic to take last modified time of files into account when identifying new and deleted files.

https://github.com/microsoft/hyperspace/pull/163/files#r489531958

### PR DESCRIPTION
**Pull request by sezruby**
### What is the context for this pull request?

 - **Tracking Issue**: #182 
 - **Parent Issue**: n/a
 - **Dependencies**: n/a

Fixes #182 

### What changes were proposed in this pull request?

Some changes from #192. (created this new PR as lack of permission to push to the branch)

This PR allows to do refresh "incremental" for modified files (same name, but different modified time or size).

`RefreshDeleteAction` handles the modified files as deleted files, and also as appended files.
`RefreshAppendAction` handles the modified files as appended files.

This PR also fixes a bug in `RefreshDeleteAction`; use the source file list of refreshed df to calculate the deleted files, instead of looking up filesystem directly. 2 different lookup points of "refreshed df" and "deleted files" can cause some correctness issue if a file is deleted between the times.

### Does this PR introduce _any_ user-facing change?

Yes, a user can refresh with "incremental" mode for modified files which was not allowed before.

### How was this patch tested?

Unit test

### CODE REVIEW

**Comment by imback82** [comment]
> `RefreshDeleteAction` handles the modified files as deleted files, and also as appended files.

Does it make sense to update `appendedFiles` / `deletedFiles` to be `Seq[FileInfo]` for completeness (esp. if someone is reading the index log entry)?

File: src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala [hunk 1]
```diff
@@ -101,60 +101,45 @@ private[actions] abstract class RefreshActionBase(
...
+    val relation = previousIndexLogEntry.relations.head
+    val originalFiles = relation.data.properties.content.fileInfos
+    val delFiles = (originalFiles -- currentFiles).map(_.name).toSeq
 
     // TODO: Add test for the scenario where existing deletedFiles and newly deleted
     //  files are updated. https://github.com/microsoft/hyperspace/issues/195.
     delFiles ++ previousIndexLogEntry.deletedFiles
   }
 
+  protected lazy val currentFiles: Set[FileInfo] = {
```
Comment (pirz) [inline src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:113]: To be consistent with other functions, please add a brief comment for the function, similar to `appendedFiles`.

**Comment by pirz** [comment]
LGTM, Thanks @sezruby

**Comment by sezruby** [comment]
> > `RefreshDeleteAction` handles the modified files as deleted files, and also as appended files.
> 
> 
> 
> Does it make sense to update `appendedFiles` / `deletedFiles` to be `Seq[FileInfo]` in index log entry for completeness (esp. if someone is reading the index log entry)?

RefreshDeleteAction only "write" the appendedFile to index log entry. So that would be ok.

**Comment by imback82** [comment]
> RefreshDeleteAction only "write" the appendedFile to index log entry. So that would be ok.

What is "ok" here? 

I know the current code works, but I am talking in terms of the metadata spec.

**Comment by sezruby** [comment]
> > RefreshDeleteAction only "write" the appendedFile to index log entry. So that would be ok.
> 
> What is "ok" here?
> 
> I know the current code works, but I am talking in terms of the metadata spec.

Oh I misread your comment. Yes Seq[FileInfo] instead of Seq[String] would be clearer and usable later.

File: src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala [hunk 2]
```diff
@@ -101,60 +101,49 @@ private[actions] abstract class RefreshActionBase(
    * are the files for which the index was never updated in the past.
    */
   protected lazy val deletedFiles: Seq[String] = {
-    val rels = previousIndexLogEntry.relations
-    val originalFiles = rels.head.data.properties.content.fileInfos
-    val currentFiles = rels.head.rootPaths
-      .flatMap { p =>
-        Content
-          .fromDirectory(path = new Path(p), throwIfNotExists = true)
```
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:109]: @imback82 BTW this change is required for this release because these lines look up filesystem directly, not from the refresh df. (though the index log entry & source.Content are generated based on the refresh df)

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:109]: OK, let me check

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:109]: The refreshed `df` also does the filessystem look up. And since both are doing it on the same `rootPaths`, wouldn't the result be the same (for non-partitioned data)?

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:109]: Yes but a file can be deleted between 2 look up point.

File: src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala [hunk 3]
```diff
@@ -207,16 +207,45 @@ class RefreshIndexTests extends QueryTest with HyperspaceSuite {
...
-      val ex = intercept[HyperspaceException](
-        hyperspace.refreshIndex(indexConfig.indexName, REFRESH_MODE_INCREMENTAL))
-      assert(
-        ex.getMessage.contains("Index refresh (to handle deleted source data) aborted. " +
-          "Existing source data file info is changed"))
+      val sourcePath = new Path(spark.read.parquet(nonPartitionedDataPath).inputFiles.head)
+      val fs = deletedFile.getFileSystem(new Configuration)
+      fs.copyToLocalFile(sourcePath, deletedFile)
+      val prevIndexLogEntry = getLatestStableLog(indexConfig.indexName)
+      assert(1 == logManager(indexConfig.indexName).getLatestId().get)
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala:215]: nit: To be consistent with other places: `assert(logManager(indexConfig.indexName).getLatestId().get == 1)`. Same for other places.

File: src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala [hunk 4]
```diff
@@ -101,60 +101,49 @@ private[actions] abstract class RefreshActionBase(
...
+   */
+  protected lazy val currentFiles: Set[FileInfo] = {
+    df.queryExecution.optimizedPlan
+      .collect {
+        case LogicalRelation(
+        HadoopFsRelation(location: PartitioningAwareFileIndex, _, _, _, _, _),
+        _,
+        _,
+        _) =>
+          location
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:125]: Can you reformat?

File: src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala (hunk 2 above)
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:109]: Got it. Could you update the PR description with this bug fix as well for the record? Thanks!

File: src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala [hunk 5]
```diff
@@ -207,16 +207,45 @@ class RefreshIndexTests extends QueryTest with HyperspaceSuite {
...
+      val indexLogEntry = getLatestStableLog(indexConfig.indexName)
+      assert(indexLogEntry.deletedFiles.isEmpty)
+      assert(indexLogEntry.appendedFiles.isEmpty)
+      assert(5 == logManager(indexConfig.indexName).getLatestId().get)
+      val files = indexLogEntry.relations.head.data.properties.content.files
+      assert(files.exists(f => f.equals(deletedFile)))
+      val v1 = indexLogEntry.content.fileInfos.count(f => f.name.contains(
+        s"${IndexConstants.INDEX_VERSION_DIRECTORY_PREFIX}=1"))
+      val v2 = indexLogEntry.content.fileInfos.count(f => f.name.contains(
+        s"${IndexConstants.INDEX_VERSION_DIRECTORY_PREFIX}=2"))
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala:246]: Can you try reformatting?

File: src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala [hunk 6]
```diff
@@ -208,16 +208,45 @@ class RefreshIndexTests extends QueryTest with HyperspaceSuite {
...
+        spark,
+        IndexLogManagerFactoryImpl.create(indexPath),
+        IndexDataManagerFactoryImpl.create(indexPath))
+        .run()
+      val indexLogEntry = getLatestStableLog(indexConfig.indexName)
+      assert(indexLogEntry.deletedFiles.isEmpty)
+      assert(indexLogEntry.appendedFiles.isEmpty)
+      assert(5 == logManager(indexConfig.indexName).getLatestId().get)
+      val files = indexLogEntry.relations.head.data.properties.content.files
+      assert(files.exists(f => f.equals(deletedFile)))
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala:242]: nit: `_.equals(...)`

**Review by imback82 — COMMENTED** [review]
Few minor comments, but generally looking good to me.

@apoorvedave1 Can you take a look at this one as well?

File: src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala (hunk 5 above)
Comment (sezruby) [reply src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala:246]: Thanks, it turns out my scalafmt didn't work properly... reformatted :)

File: src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala [hunk 7]
```diff
@@ -208,16 +208,46 @@ class RefreshIndexTests extends QueryTest with HyperspaceSuite {
...
+      new RefreshDeleteAction(
+        spark,
+        IndexLogManagerFactoryImpl.create(indexPath),
+        IndexDataManagerFactoryImpl.create(indexPath))
+        .run()
+
+      val indexLogEntryAfterDeleteRefresh = getLatestStableLog(indexConfig.indexName)
+      assert(logManager(systemPath, indexConfig.indexName).getLatestId().get == 3)
+      assert(indexLogEntryAfterDeleteRefresh.deletedFiles.isEmpty)
+      assert(indexLogEntryAfterDeleteRefresh.appendedFiles.size == 1)
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala:230]: According to the PR description below, the modified file should be recorded as "deletedFiles"?
```
RefreshDeleteAction handles the modified files as deleted files, and also as appended files.
```

File: src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala [hunk 8]
```diff
@@ -101,60 +101,49 @@ private[actions] abstract class RefreshActionBase(
...
-        case None => delFiles :+= f.name
-      }
-    }
+    val relation = previousIndexLogEntry.relations.head
+    val originalFiles = relation.data.properties.content.fileInfos
+    val delFiles = (originalFiles -- currentFiles).map(_.name).toSeq
 
     // TODO: Add test for the scenario where existing deletedFiles and newly deleted
     //  files are updated. https://github.com/microsoft/hyperspace/issues/195.
     delFiles ++ previousIndexLogEntry.deletedFiles
```
[the remark this answers is not in the document]
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:110]: This can be addressed as a follow-up PR if needed.

**Comment by imback82** [comment]
> > > RefreshDeleteAction only "write" the appendedFile to index log entry. So that would be ok.
> > 
> > 
> > What is "ok" here?
> > I know the current code works, but I am talking in terms of the metadata spec.
> 
> Oh I misread your comment. Yes Seq[FileInfo] instead of Seq[String] would be clearer and usable later.

Thanks, created #210.

File: src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:110]: Created #211 if we need to follow up on this.

File: src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala (hunk 7 above)
Comment (imback82) [reply src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala:230]: Ah, nm, I think we need the test coverage for `RefreshAppendAction` where `deletedFiles` become non-empty.

Comment (imback82) [reply src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala:230]: @sezruby can you do a follow up PR to add the coverage? (I manually verified by updating the test slightly to unblock the current release.)

**Review by imback82 — APPROVED** [review]
LGTM (few follow up items), thanks @sezruby!

File: src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala (hunk 7 above)
Comment (sezruby) [reply src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala:230]: Created #213.
